Python/specialize.c

```diff
@@ -1195,3 +1199,32 @@ _Py_Specialize_BinaryAdd(PyObject *left, PyObject *right, _Py_CODEUNIT *instr)
     assert(!PyErr_Occurred());
     return 0;
 }
+
+int
+_Py_Specialize_BinaryMultiply(PyObject *left, PyObject *right, _Py_CODEUNIT *instr)
+{
+    if (!Py_IS_TYPE(left, Py_TYPE(right))) {
+        SPECIALIZATION_FAIL(BINARY_MULTIPLY, SPEC_FAIL_DIFFERENT_TYPES);
+        goto fail;
+    }
+    if (PyLong_CheckExact(left)) {
+        *instr = _Py_MAKECODEUNIT(BINARY_MULTIPLY_INT, saturating_start());
+        goto success;
+    }
+    else if (PyFloat_CheckExact(left)) {
+        *instr = _Py_MAKECODEUNIT(BINARY_MULTIPLY_FLOAT, saturating_start());
+        goto success;
+    }
+    else {
+        SPECIALIZATION_FAIL(BINARY_MULTIPLY, SPEC_FAIL_OTHER);
```

Comment:
Maybe this should be refined further. In another PR, though.

Reply:
I agree. I don't think it will touch pyperformance, but it seems that in our own test suite almost 10% aren't hits https://github.com/python/cpython/pull/28727#issuecomment-934067784.

This probably varies wildly, I'd imagine ``test_string`` has lots of ``str + str`` too.

Reply:
Most of the non-hits are deferred, not specialization failure, so I think that's because the nature of a lot of test code is to only be run once, without many tight loops.

Reply:
Thanks for pointing that out Dennis. I can't believe I missed that :).

Off-topic: I've recently wondered if ``pyperformance`` is somewhat out of touch (no offence intended to its contributors). Many of its benchmarks (nbody, nqueens, etc.) have been found by the JS folks to not be realistic. The Pyston benchmark suite seems way more comprehensive.

Reply:
"deferred" means that the `ADAPTIVE` instruction is counting down until its next specialization attempt.
This only happens *after* a specialization failure.

Look at the numbers. `deferred ≈ `sum(specialization-failures) * ADAPTIVE_CACHE_BACKOFF`